pkg/queue/workers/task_worker.go

```diff
@@ -34,52 +30,116 @@ type taskWorker struct {
 
 	handler  queue.TaskHandler
 	dequeuer queue.Dequeuer
-	maxWait  time.Duration
+	interval time.Duration
 
 	queue.Worker
 }
 
-func (w *taskWorker) iteration(ctx context.Context, tracer opentracing.Tracer, ticker *time.Ticker) (err error) {
-	logrus.Debug("starting work attempt")
+func (w *taskWorker) Work(ctx context.Context) (err error) {
+	tracer := opentracing.GlobalTracer()
+	ticker := time.NewTicker(w.interval)
+	defer ticker.Stop()
+	queue.TaskWorkerMetrics.ActiveGauge.Inc()
+	defer queue.TaskWorkerMetrics.ActiveGauge.Dec()
+
+	// the error in the iteration should not stop the work
+	// it's logged by the Tracer interface, so we don't have to handle it here
+	// since the ticker delivers the first tick after the interval we need to run it for the
+	// first time out of the loop
+	e := w.iteration(ctx, tracer)
+	if e != nil {
+		logrus.Error(e)
+	}
 
+	logrus.Debug("starting task worker loop...")
+	// while ctx is not cancelled or interrupted
+	for {
+		queue.TaskWorkerMetrics.WaitingGauge.Inc()
+
+		select {
+		case <-ctx.Done():
+			queue.TaskWorkerMetrics.WaitingGauge.Dec()
+			logrus.Debug("processing loop is interrupted")
+			return ctx.Err()
+		case <-ticker.C:
+			queue.TaskWorkerMetrics.WaitingGauge.Dec()
+			e := w.iteration(ctx, tracer)
```

Comment:
this seems to be different from the previous code.  Previously, the ticker would cancel the current iteration (via the context cancel) and start a new one. Here the ticker clearly starts a new iteration, but how does it cancel the previous iteration?

This was achieved before by creating a new `ContextWithCancel` from the parent work `ctx` for each iteration attempt and then at the end of the tick, calling the cancel.  Here in this implementation all of the iterations share the same context instead of children of the parent work context.

Reply:
This is now consistent with the scheduler and it's completely different logic, what it does now:

1. try fetch a task
  1.2. if `sql.ErrNoRows` then wait the interval and go to 1. after that
  1.3. if there is task handle it and go to 1.

this allows to handle several tasks from the queue in one iteration without waiting extra time, that's what I meant by "making the task worker more efficient".

The previous code was over-complicated and I still don't understand some parts of it, the current algorithm has been working in scheduler for a long time and battle-tested.

I tested it with Hub and datastore and didn't notice any difference in the behavior, here is how tracing looks like:

![Screenshot from 2020-09-26 12-26-43](https://user-images.githubusercontent.com/887952/94338558-8f07a480-fff3-11ea-81ed-d6ab8e5340be.png)

Reply:
I take it back, these are all scheduler's spans, for some reason it works differently for the task worker:
![Screenshot from 2020-09-26 12-29-42](https://user-images.githubusercontent.com/887952/94338611-076e6580-fff4-11ea-95c2-0e92d52c6bab.png)

Reply:
I think the difference is that the scheduler skips locked rows and the task worker does not.

Reply:
I added the timeout and now the task worker traces look like this
![Screenshot from 2020-09-26 12-59-26](https://user-images.githubusercontent.com/887952/94339169-2a9b1400-fff8-11ea-884e-5f621fa62ec6.png)